src/core/manager.c

```diff
@@ -3771,6 +3829,13 @@ int manager_reload(Manager *m) {
          * and everything else that is worth flushing out. We'll get it all back from the serialization — if we need
          * it.*/
 
+        /* We need existing BPF programs to survive reload, otherwise there will be a period where no BPF
+         * program is active during task execution within a cgroup. This would be bad since this may have
+         * security or reliability implications: devices we should filter won't be filtered, network activity
+         * we should filter won't be filtered, etc. We pin all the existing devices by bumping their
+         * refcount, and then storing them to later have it decremented. */
+        (void) manager_pin_all_cgroup_bpf_programs(m, pinned_progs);
```

Comment:
Shouldn't this be `&pinned_progs`?  Otherwise how can this function possibly allocate the set and place it @ pinned_progs in this scope?

Reply:
This is already a `Set *`, why would it need to be `&pinned_progs`?

Reply:
Perhaps I'm still waking up here, but pinned_progs is initialized to NULL here.

You're passing NULL to manager_pin_all_cgroup_bpf_programs().  How can that function allocate the Set and store its pointer anywhere accessible by this outer function when all it's receiving is NULL?

Reply:
Huh, now I see what you mean and wonder why it works anyway.

```
[root@ktst ~]# journalctl -b | grep -i pinned
Oct 30 16:01:00 ktst systemd[1]: Pinned 20 BPF programs for daemon-reload
Oct 30 16:01:02 ktst systemd[1]: Pinned 20 BPF programs for daemon-reload
Oct 30 17:27:47 ktst systemd[1]: Pinned 9 BPF programs for daemon-reload
Oct 30 17:57:02 ktst systemd[1]: Pinned 9 BPF programs for daemon-reload
```

Reply:
It probably worked because you supply a pointer to the local on-stack pinned_progs parameter in manager_pin_all_cgroup_bpf_programs().  So it didn't dereference NULL, but it only stored the Set local to the function.

The later unpin would just see NULL and unpin nothing.